File: chazm-api/src/main/java/runtimemodels/chazm/api/entity/InstanceGoal.java
package runtimemodels.chazm.api.entity;

import runtimemodels.chazm.api.Organization;
import runtimemodels.chazm.api.id.Identifiable;

import java.util.Map;

/**
 * The {@linkplain InstanceGoal} interface defines the instance goal, which is a concrete instantiation of a {@linkplain SpecificationGoal}, of an
 * {@linkplain Organization}.
 *
 * @author Christopher Zhong
 * @since 3.4
 */
public interface InstanceGoal extends Identifiable<InstanceGoal> {
    /**
     * Returns the {@linkplain SpecificationGoal} that instantiated this {@linkplain InstanceGoal}.
     *
     * @return the {@linkplain SpecificationGoal} that instantiated this {@linkplain InstanceGoal}.
     */
    SpecificationGoal getGoal();

    /**
     * Returns a {@linkplain Map} that represents the parameters of this {@linkplain InstanceGoal}.
     *
     * @return a {@linkplain Map} that represents the parameters of this {@linkplain InstanceGoal}.
     */
    Map<Object, Object> getParameters();

    /**
     * Adds the given {@linkplain Object} {@code key} and {@linkplain Object} {@code value} as a parameter to this {@linkplain InstanceGoal}.
     *
     * @param key   an {@linkplain Object} representing the {@code key}.
     * @param value an {@linkplain Object} representing the {@code value}.
     */
    void addParameter(Object key, Object value);

    /**
     * Removes a parameter by the given {@linkplain Object} {@code key} from this {@linkplain InstanceGoal}.
     *
     * @param key an {@linkplain Object} representing the {@code key}.
     */
    void removeParameter(Object key);
}

File: chazm-api/src/main/kotlin/runtimemodels/chazm/api/entity/InstanceGoal.kt
package runtimemodels.chazm.api.entity

import runtimemodels.chazm.api.id.Identifiable
import runtimemodels.chazm.api.id.InstanceGoalId
import runtimemodels.chazm.api.organization.Organization

/**
 * The [InstanceGoal] interface defines an instance goal, which is a concrete instantiation
 * of a [SpecificationGoal], of an [Organization].
 *
 * @author Christopher Zhong
 * @see [SpecificationGoal]
 * @since 3.4
 */
interface InstanceGoal : Identifiable<InstanceGoal, InstanceGoalId> {
    /**
     * The [GoalParameter] interface defines a wrapper for a [Map] of [Any].
     */
    interface GoalParameter : Map<Any, Any> {
        /**
         * Adds the given `key` and `value` as a parameter to this [InstanceGoal].
         *
         * @param key   an [Any] representing the `key`.
         * @param value an [Any] representing the `value`.
         */
        fun add(key: Any, value: Any)

        /**
         * Removes a parameter by the given `key` from this [InstanceGoal].
         *
         * @param key an [Any] representing the `key`.
         */
        fun remove(key: Any)
    }

    /**
     * The [SpecificationGoal] that represents this [InstanceGoal].
     */
    val goal: SpecificationGoal

    /**
     * A [GoalParameter] that represents the parameters of this [InstanceGoal].
     */
    val parameters: GoalParameter
}
